com.ca.lsp.cobol/lsp-service-cobol/src/main/java/com/ca/lsp/cobol/service/CopybookServiceImpl.java

```diff
@@ -135,120 +115,47 @@ private Path retrievePathOrNull(String filename, List<Path> datasetPathList) {
   }
 
   private List<Path> generatePathListFromSettings(String profile, List<String> datasetList) {
-    // can happen here that copybooks or internal structure is null
     return datasetList.stream()
         .map(
             it ->
                 Paths.get(
-                    getCopybookFolder(getWorkspaceFoldersAsPathList().get(0))
+                    getCopybookBaseFolder(getWorkspaceFoldersAsPathList().get(0))
                         + filesystemSeparator()
                         + profile
                         + filesystemSeparator()
                         + it))
-        .filter(Objects::nonNull)
         .collect(Collectors.toList());
   }
 
-  /**
-   * @param fileName copybook name
-   * @param targetFolderPath physical path of workspace where to search for the copybook
-   * @return Path of the found copybook in the target folder.
-   */
-  private Path applySearch(String fileName, Path targetFolderPath) {
-    try (Stream<Path> pathStream =
-        Files.find(
-            targetFolderPath,
-            100,
-            (path, basicFileAttributes) -> isValidFileFound(path.toFile(), fileName),
-            FileVisitOption.FOLLOW_LINKS)) {
-      return pathStream.findAny().orElse(null);
-    } catch (IOException e) {
-      log.error(e.getMessage());
-      return null;
-    }
-  }
-
   private List<Path> getWorkspaceFoldersAsPathList() {
     return Optional.ofNullable(getWorkspaceFolders())
```

Comment:
Do we assume the workspace folders as static? If yes, we may save the result of `getWorkspaceFoldersAsPathList()` instead of storing the `workspaceFolders`. It will allow us not to repeat this conversion every time.

Reply:
Is allowed to change the value of ```WorkspaceFolders``` cause the user can open a new workspace and at the initialize handshake we'll get the new value of that.
I solved in a way that when ```setWorkspaceFolders()``` is triggered, we calculate the ```List<Path>``` from that and store it internally. No one will calculate again this list (until you close and open a new workspace in the IDE)

Reply:
If I understand correctly, we will restart the server if the user does this due to the client also restarts. If yes, then from our point of view the workspace folders are static and may be set at the initialization.

Reply:
But if you close a workspace folder the server will go in shutdown? If yes what you said is correct but more in general I prefer don't rely on static things, we control when the List<Paths> is generated and I think it should be enough robust - is my opinion of course.

Reply:
Static here is not from the Java point of view, it is more about that this won't change during the application lifecycle

Reply:
I'd like to postpone this discussion when I will commit the changes, I found a violation of responsibility in the ```CopybookDependencyServiceImpl``` and I'm working hard to refactor this thing.